File: src/main/java/net/lomeli/trophyslots/KotlinAdapter.java
package net.lomeli.trophyslots;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

import net.minecraftforge.fml.common.FMLModContainer;
import net.minecraftforge.fml.common.ILanguageAdapter;
import net.minecraftforge.fml.common.ModContainer;
import net.minecraftforge.fml.relauncher.Side;

public class KotlinAdapter implements ILanguageAdapter {

    private Logger log = LogManager.getLogger("ILanguageAdapter/Kotlin");

    @Override
    public Object getNewInstance(FMLModContainer container, Class<?> objectClass, ClassLoader classLoader, Method factoryMarkedAnnotation) throws Exception {
        log.debug("FML has asked for {} to be constructed...", objectClass.getSimpleName());
        try {
            // Try looking for an object type
            Field f = objectClass.getField("INSTANCE$");
            Object obj = f.get(null);
            if (obj == null) throw new NullPointerException();
            log.debug("Found an object INSTANCE$ reference in {}, using that. ({})", objectClass.getSimpleName(), obj);
            return obj;
        } catch (Exception ex) {
            // Try looking for a class type
            log.debug("Failed to get object reference, trying class construction.");
            try {
                Object obj = objectClass.newInstance();
                if (obj == null) throw new NullPointerException();
                log.debug("Constructed an object from a class type ({}), using that. ({})", objectClass, obj);
                return obj;
            } catch (Exception e) {
                throw new KotlinAdapterException(e);
            }
        }
    }

    @Override
    public boolean supportsStatics() {
        return false;
    }

    @Override
    public void setProxy(Field target, Class<?> proxyTarget, Object proxy) throws IllegalArgumentException, IllegalAccessException, NoSuchFieldException, SecurityException {
        log.debug("Setting proxy: {}.{} -> {}", target.getDeclaringClass().getSimpleName(), target.getName(), proxy);
        for (Field x : proxyTarget.getFields()) {
            if (x.getName().equals("INSTANCE$")) {
                try {
                    log.debug("Setting proxy on INSTANCE$; singleton target.");
                    Object obj = proxyTarget.getField("INSTANCE$").get(null);
                    target.set(obj, proxy);
                } catch (Exception ex) {
                    throw new KotlinAdapterException(ex);
                }
            } else
                target.set(proxyTarget, proxy);
        }
    }

    @Override
    public void setInternalProxies(ModContainer mod, Side side, ClassLoader loader) {

    }

    private class KotlinAdapterException extends RuntimeException {
        public KotlinAdapterException(Exception ex) {
            super("Kotlin adapter error - do not report to Forge!", ex);
        }
    }
}

File: src/main/kotlin/net/lomeli/trophyslots/KotlinAdapter.kt
package net.lomeli.trophyslots

import net.minecraftforge.fml.common.FMLModContainer
import net.minecraftforge.fml.common.ILanguageAdapter
import net.minecraftforge.fml.common.ModContainer
import net.minecraftforge.fml.relauncher.Side
import org.apache.logging.log4j.LogManager
import java.lang.reflect.Field
import java.lang.reflect.Method

/**
 * Kotlin implementation of FML's ILanguageAdapter.
 *
 * Use by setting <pre>modLanguageAdapter = "io.drakon.forgelin.KotlinAdapter"</pre> in the Mod annotation
 * (Forge 1.8-11.14.1.1371 or above required).
 *
 * @author Arkan <arkan@drakon.io>
 */
public class KotlinAdapter : ILanguageAdapter {

    companion object metadata {
        public final val ADAPTER_VERSION: String = "@VERSION@-@KOTLIN@"
    }

    private val log = LogManager.getLogger("ILanguageAdapter/Kotlin")
    override fun supportsStatics(): Boolean {
        return false
    }

    override fun setProxy(target: Field?, proxyTarget: Class<*>?, proxy: Any?) {
        if (target != null && proxyTarget != null) {
            log.debug("Setting proxy: {}.{} -> {}", target.declaringClass.simpleName, target.name, proxy)
            if (proxyTarget.fields.any() { x -> x.name.equals("INSTANCE$") }) {
                // Singleton
                try {
                    log.debug("Setting proxy on INSTANCE$; singleton target.")
                    val obj = proxyTarget.getField("INSTANCE$").get(null)
                    target.set(obj, proxy)
                } catch (ex: Exception) {
                    throw KotlinAdapterException(ex)
                }
            } else {
                //TODO Log?
                target.set(proxyTarget, proxy)
            }
        }
    }

    override fun getNewInstance(container: FMLModContainer?, objectClass: Class<*>?, classLoader: ClassLoader?, factoryMarkedAnnotation: Method?): Any? {
        if (objectClass != null) {
            log.debug("FML has asked for {} to be constructed...", objectClass.simpleName)
            try {
                // Try looking for an object type
                val f = objectClass.getField("INSTANCE$")
                val obj = f.get(null)
                if (obj == null) throw NullPointerException()
                log.debug("Found an object INSTANCE$ reference in {}, using that. ({})", objectClass.simpleName, obj)
                return obj
            } catch (ex: Exception) {
                // Try looking for a class type
                log.debug("Failed to get object reference, trying class construction.")
                try {
                    val obj = objectClass.newInstance()
                    if (obj == null) throw NullPointerException()
                    log.debug("Constructed an object from a class type ({}), using that. ({})", objectClass, obj)
                    log.warn("Hey, you, modder who owns {} - you should be using 'object' instead of 'class' on your @Mod class.", objectClass.simpleName)
                    return obj
                } catch (ex: Exception) {
                    throw KotlinAdapterException(ex)
                }
            }
        }
        return null;
    }

    override fun setInternalProxies(mod: ModContainer?, side: Side?, loader: ClassLoader?) {
        // Nothing to do; FML's got this covered for Kotlin.
    }

    private class KotlinAdapterException(ex: Exception) : RuntimeException("Kotlin adapter error - do not report to Forge!", ex)

}